exchange/bitswap/decision/engine.go

```diff
@@ -91,45 +93,55 @@ func NewEngine(ctx context.Context, bs bstore.Blockstore) *Engine {
 }
 
 func (e *Engine) taskWorker(ctx context.Context) {
-	log := log.Prefix("bitswap.Engine.taskWorker")
+	defer close(e.outbox) // because taskWorker uses the channel exclusively
+	for {
+		oneTimeUse := make(chan Envelope, 1) // buffer to prevent blocking
+		select {
+		case <-ctx.Done():
+			return
+		case e.outbox <- oneTimeUse:
+		}
+		// receiver is ready for an outoing envelope. let's prepare one. first,
+		// we must acquire a task from the PQ...
+		envelope, err := e.nextEnvelope(ctx)
+		if err != nil {
+			close(oneTimeUse)
+			return // ctx cancelled
+		}
+		oneTimeUse <- *envelope // buffered. won't block
```

Comment:
personally, I would make the `oneTimeUse` be a `chan *Envelope` just to avoid a dereference, looks cleaner that way IMO